### PR TITLE
Use iptables-restore to preserve order with MASQ & FORWARD rules

### PR DESCRIPTION
**Pull request by ryarnyah**
Use iptables-restore to preserve order with MASQ & FORWARD rules.

## Description
When anything happen when delete/append iptables rules, sometimes you can have your rules out of order.This PR use only iptables-restore to guaranty order in delete/append rules.

This close https://github.com/coreos/flannel/issues/1261 & https://github.com/coreos/flannel/issues/1146

## Todos
- [X] Tests
- [x] Documentation (Needed?)
- [X] Release note

## Release Note

```release-note
None required
```

### CODE REVIEW

**Comment by rajatchopra** [comment]
Looks good. How can we have an end-to-end test for it?
Maybe check on iptables rules after doing a kube e2e in dist/function-test-k8s.sh?

**Comment by ryarnyah** [comment]
e2e tests has been added as requested

**Comment by ryarnyah** [comment]
Do you need something else to have this PR merged?

**Comment by luthermonson** [comment]
@Oats87 take a look would you?

**Comment by luthermonson** [comment]
@Oats87 sign off on this so we can merge it, youre the iptables ninja

**Comment by Oats87** [comment]
@ryarnyah would you mind rebasing this PR (or just fixing the merge conflict)?

**Comment by ryarnyah** [comment]
@Oats87 i will look to rebase it this weekend if i have time ;)

**Comment by ryarnyah** [comment]
@Oats87 It's seem to be good for review now.

File: network/iptables.go [hunk 1]
```diff
@@ -100,21 +99,74 @@ func ipTablesRulesExist(ipt IPTables, rules []IPTablesRule) (bool, error) {
...
+	}
+
+	log.V(6).Infof("trying to run iptable-restore < %+v", tablesRules)
+
+	err = iptRestore.ApplyPartial(tablesRules)
+	if err != nil {
+		return fmt.Errorf("failed to apply partial iptables-restore %v", err)
+	}
+
+	log.Infof("bootstrap done")
```
Comment (Oats87) [inline network/iptables.go:138]: Can we provide some more context in this error message? Does this have to be done at the info level?

File: network/iptables.go [hunk 2]
```diff
@@ -130,11 +182,21 @@ func DeleteIPTables(rules []IPTablesRule) error {
 		log.Errorf("Failed to setup IPTables. iptables binary was not found: %v", err)
 		return err
 	}
-	teardownIPTables(ipt, rules)
+	iptRestore, err := NewIPTablesRestore()
+	if err != nil {
+		// if we can't find iptables, give up and return
+		log.Errorf("Failed to setup IPTables-restore: %v", err)
```
Comment (Oats87) [inline network/iptables.go:188]: nit: this should probably be `iptables-restore` instead of `IPTables-restore`

File: dist/functional-test-k8s.sh [hunk 3]
```diff
@@ -83,7 +83,7 @@ start_flannel() {
        docker run -e NODE_NAME=flannel$host_num --privileged --name flannel-e2e-test-flannel$host_num -id --entrypoint /bin/sh $FLANNEL_DOCKER_IMAGE >/dev/null
        docker exec flannel-e2e-test-flannel$host_num /bin/sh -c 'mkdir -p /etc/kube-flannel'
        echo $flannel_conf | docker exec -i flannel-e2e-test-flannel$host_num /bin/sh -c 'cat > /etc/kube-flannel/net-conf.json'
-       docker exec -d flannel-e2e-test-flannel$host_num /opt/bin/flanneld --kube-subnet-mgr --kube-api-url $k8s_endpt
+       docker exec -d flannel-e2e-test-flannel$host_num /opt/bin/flanneld --kube-subnet-mgr --ip-masq --kube-api-url $k8s_endpt
```
Comment (Oats87) [inline dist/functional-test-k8s.sh:86]: Do we actually need this `--ip-masq`?

File: dist/functional-test-k8s.sh [hunk 4]
```diff
@@ -148,6 +152,37 @@ pings() {
 	assert "docker exec --privileged flannel-e2e-test-flannel2 /bin/ping -c 5 $ping_dest1" "Host 2 cannot ping host 1"
 }
 
+check_iptables() {
```
Comment (Oats87) [inline dist/functional-test-k8s.sh:155]: Should we maybe have a specific test where we explicitly garble rule order to ensure that `flanneld` corrects the ordering of the rules?

File: network/iptables.go [hunk 5]
```diff
@@ -100,21 +99,74 @@ func ipTablesRulesExist(ipt IPTables, rules []IPTablesRule) (bool, error) {
 	return true, nil
 }
 
+func ipTablesCleanAndBuild(ipt IPTables, rules []IPTablesRule) (IPTablesRestoreRules, error) {
+	tablesRules := IPTablesRestoreRules{}
+
+	// Build delete rules
```
Comment (Oats87) [inline network/iptables.go:105]: Is there any way to ensure the order is wrong before just automatically deleting/creating the rules?

File: network/iptables.go [hunk 6]
```diff
@@ -100,21 +99,74 @@ func ipTablesRulesExist(ipt IPTables, rules []IPTablesRule) (bool, error) {
...
+	}
+
+	return tablesRules, nil
+}
+
+func ipTablesBootstrap(ipt IPTables, iptRestore IPTablesRestore, rules []IPTablesRule) error {
+	tablesRules, err := ipTablesCleanAndBuild(ipt, rules)
+	if err != nil {
+		// if we can't find iptables, give up and return
+		return fmt.Errorf("Failed to setup IPTables-restore payload: %v", err)
```
Comment (Oats87) [inline network/iptables.go:128]: nit: should be `iptables-restore` not `IPTables-restore`

File: network/iptables.go [hunk 7]
```diff
@@ -100,21 +99,74 @@ func ipTablesRulesExist(ipt IPTables, rules []IPTablesRule) (bool, error) {
...
+	}
+
+	log.V(6).Infof("trying to run iptable-restore < %+v", tablesRules)
```
Comment (Oats87) [inline network/iptables.go:131]: nit: `iptables-restore` not `iptable-restore`

File: network/iptables.go [hunk 8]
```diff
@@ -100,21 +99,74 @@ func ipTablesRulesExist(ipt IPTables, rules []IPTablesRule) (bool, error) {
...
 	ipt, err := iptables.New()
 	if err != nil {
 		// if we can't find iptables, give up and return
 		log.Errorf("Failed to setup IPTables. iptables binary was not found: %v", err)
 		return
 	}
+	iptRestore, err := NewIPTablesRestore()
+	if err != nil {
+		// if we can't find iptables, give up and return
+		log.Errorf("Failed to setup IPTables. iptables binary was not found: %v", err)
```
Comment (Oats87) [inline network/iptables.go:153]: Isn't the problem that the `iptables-restore` binary was not found?

File: network/iptables_restore.go [hunk 9]
```diff
@@ -0,0 +1,213 @@
+// Copyright 2015 flannel authors
```
Comment (Oats87) [inline network/iptables_restore.go:1]: This probably actually belongs in `coreos/go-iptables` but having it here should be fine.

File: network/iptables_restore.go [hunk 10]
```diff
@@ -0,0 +1,213 @@
...
+	ipTablesRestoreCmd  string = "iptables-restore"
+	ip6TablesRestoreCmd string = "ip6tables-restore"
+)
+
+// IPTablesRestore wrapper for iptables-restore
+type IPTablesRestore interface {
+	// ApplyFully apply all rules and flush chains
+	ApplyFully(rules IPTablesRestoreRules) error
+	// ApplyPartial apply without flush chains
+	ApplyPartial(rules IPTablesRestoreRules) error
```
Comment (Oats87) [inline network/iptables_restore.go:41]: Maybe call this `ApplyNoFlush` or `ApplyWithoutFlush`?

**Review by Oats87 — CHANGES_REQUESTED** [review]

File: network/iptables.go (hunk 6 above)
Comment (briandowns) [reply network/iptables.go:128]: nit: error messages shouldn't be capitalized

**Comment by louiznk** [comment]
> @ryarnyah could you rebase and respond to the review comments?

Hello @rajatchopra if you and @ryarnyah are agree I can do this for @ryarnyah 
Thanks

**Comment by luthermonson** [comment]
@louiznk please do

File: network/iptables_restore.go [hunk 11]
```diff
@@ -0,0 +1,213 @@
...
+}
+
+// ipTablesRestore internal type
+type ipTablesRestore struct {
+	path     string
+	protocol iptables.Protocol
```
[the remark this answers is not in the document]
Comment (louiznk) [reply network/iptables_restore.go:48]: :ok_hand:

**Comment by luthermonson** [comment]
@louiznk any chance you can answer/fix for @Oats87 feedback? can probably wrap this up and merge if those are addressed

File: network/iptables_restore.go [hunk 12]
```diff
@@ -0,0 +1,213 @@
...
+// IPTablesRestore wrapper for iptables-restore
+type IPTablesRestore interface {
+	// ApplyFully apply all rules and flush chains
+	ApplyFully(rules IPTablesRestoreRules) error
+	// ApplyPartial apply without flush chains
+	ApplyPartial(rules IPTablesRestoreRules) error
+}
+
+// ipTablesRestore internal type
+type ipTablesRestore struct {
```
[the remark this answers is not in the document]
Comment (louiznk) [reply network/iptables_restore.go:46]: @Oats87 I'm not sur to understood, can you give me more details ?

File: network/iptables_restore.go [hunk 13]
```diff
@@ -0,0 +1,213 @@
...
+		return false, err
+	}
+	v1, v2, v3, err := extractIptablesRestoreVersion(version)
+	if err != nil {
+		return false, err
+	}
+	return ipTablesHasWaitSupport(v1, v2, v3), nil
+}
+
+// Checks if an iptables-restore version is after 1.4.20, when --wait was added
```
[the remark this answers is not in the document]
Comment (louiznk) [reply network/iptables_restore.go:156]: :ok_hand:

File: network/iptables.go (hunk 1 above)
Comment (louiznk) [reply network/iptables.go:173]: I'm not sure it's possible

File: network/iptables.go (hunk 2 above)
Comment (louiznk) [reply network/iptables.go:257]: :ok_hand:

File: network/iptables.go (hunk 6 above)
Comment (louiznk) [reply network/iptables.go:163]: :ok_hand:

File: network/iptables.go (hunk 7 above)
Comment (louiznk) [reply network/iptables.go:166]: :ok_hand:

File: network/iptables.go (hunk 8 above)
Comment (louiznk) [reply network/iptables.go:188]: :ok_hand:

File: network/iptables_restore.go (hunk 10 above)
Comment (louiznk) [reply network/iptables_restore.go:42]: :ok:

File: dist/functional-test-k8s.sh (hunk 3 above)
Comment (louiznk) [reply dist/functional-test-k8s.sh:87]: Without this somes tests fails (test_hors-gw, test_ipip, test_udp, test_vxlan).

File: network/iptables.go (hunk 5 above)
Comment (louiznk) [reply network/iptables.go:140]: I think it's possible if we have before memorized the order, but I'm not sure we can in every case memorize the order. And we will need to store this state outside the memory (ie: etcd) for the crash case. 
So I'm not sure it's a real benefit compared to the complexity it's will add.

File: dist/functional-test-k8s.sh (hunk 4 above)
Comment (louiznk) [reply dist/functional-test-k8s.sh:162]: I need to check how can I do this. Do you have an idea ?

File: network/iptables_restore.go (hunk 9 above)
Comment (louiznk) [reply network/iptables_restore.go:1]: So I leave it that way ?

**Comment by louiznk** [comment]
> was requested for review

Hello @luthermonson 
I don't know if you have the opportunity to see the commit and question/reply for @Oats87 feedback?
Thanks a lot for your help and review.

**Review by louiznk — COMMENTED** [review]
Apply requests

**Review by louiznk — APPROVED** [review]

**Comment by manuelbuil** [comment]
Are you still interested in merging this? Could you rebase so that we can review it again? Thanks!

**Comment by louiznk** [comment]
Hello,

Thanks, I can do that next weekend.


Le jeu. 26 mai 2022 à 17:57, Manuel Buil ***@***.***> a
écrit :

**Comment by louiznk** [comment]
> Are you still interested in merging this? Could you rebase so that we can review it again? Thanks!

Hello @manuelbuil ,

Thanks, I can do that this week end

**Comment by louiznk** [comment]
> Are you still interested in merging this? Could you rebase so that we can review it again? Thanks!

Hello @manuelbuil , the rebase is done, I fixe some trouble with this new versions (usage of comments on iptables rules for exemples)
Thanks for your review and feedback

File: dist/functional-test-k8s.sh (hunk 4 above)
Comment (louiznk) [reply dist/functional-test-k8s.sh:175]: As I know with iptables-restore the order is garanty in the transaction and the `--wait` option prevent multiple instances of the program from running concurrently (so no concurrently changes on the iptables rules)

**Review by manuelbuil — COMMENTED** [review]
IIUC, some users detected a race condition in which these rules: https://github.com/flannel-io/flannel/blob/master/network/iptables.go#L57-L66, are applied in different order. In my opinion, the simplest solution would be adding a lock here https://github.com/flannel-io/flannel/blob/master/network/iptables.go#L30 and then check the lock around here: https://github.com/flannel-io/flannel/blob/master/network/iptables.go#L244 and not allow any other rule to be added until the lock is released. That way the race condition will not exist anymore and the problem would be fixed, or? Using iptables-restore also fixes the problem but it requires adding ~500 lines of code, which makes me wonder if this is an overkill given that the use of iptables in flannel is minimal (if we had some sort of network policy engine using iptables, I would probably think differently)

File: network/iptables_restore.go [hunk 14]
```diff
@@ -0,0 +1,226 @@
...
+
+// IPTablesRestoreRules represents iptables-restore table block
+type IPTablesRestoreRules map[string][]IPTablesRestoreRuleSpec
+
+// IPTablesRestoreRuleSpec represents one rule spec delimited by space
+type IPTablesRestoreRuleSpec []string
+
+// NewIPTablesRestore build new IPTablesRestore for IPv4
+func NewIPTablesRestore() (IPTablesRestore, error) {
+	return NewIPTablesRestoreWithProtocol(iptables.ProtocolIPv4)
```
[the remark this answers is not in the document]
Comment (louiznk) [reply network/iptables_restore.go:58]: > IIUC, some users detected a race condition in which these rules: https://github.com/flannel-io/flannel/blob/master/network/iptables.go#L57-L66, are applied in different order. In my opinion, the simplest solution would be adding a lock here https://github.com/flannel-io/flannel/blob/master/network/iptables.go#L30 and then check the lock around here: https://github.com/flannel-io/flannel/blob/master/network/iptables.go#L244 and not allow any other rule to be added until the lock is released. That way the race condition will not exist anymore and the problem would be fixed, or? Using iptables-restore also fixes the problem but it requires adding ~500 lines of code, which makes me wonder if this is an overkill given that the use of iptables in flannel is minimal (if we had some sort of network policy engine using iptables, I would probably think differently).
> 
> I'd love to hear more opinions!
You're right in the case of flannel alone. But I have trouble with some concurrent access from other programs (in production).
In my case it was kube-router (using for network policy).
I know there are better solutions than using flannel + kube-routeur, but unfortunately not in special my case (some external requirements).
So the main advantage for me with iptables-restore is the "transaction" that allows to preserve the order when you have some concurrents access to iptables.

[the remark this answers is not in the document]
Comment (louiznk) [reply network/iptables_restore.go:58]: > Is there a reason to hardcode IPv4?

Just for respecting the same convention as implementation based on iptables (New => ipv4 by default and NewWithProtocol => protocol in parameter). For exemple see : 
https://github.com/flannel-io/flannel/blob/master/network/iptables.go#L190 and https://github.com/flannel-io/flannel/blob/master/network/iptables.go#L206
I can change it if it's better

File: network/iptables.go [hunk 15]
```diff
@@ -135,24 +134,77 @@ func ipTablesRulesExist(ipt IPTables, rules []IPTablesRule) (bool, error) {
...
+			tablesRules[rule.table] = append(tablesRules[rule.table], append(IPTablesRestoreRuleSpec{"-D", rule.chain}, rule.rulespec...))
+		}
+		tablesRules[rule.table] = append(tablesRules[rule.table], append(IPTablesRestoreRuleSpec{"-A", rule.chain}, rule.rulespec...))
+	}
+
+	return tablesRules, nil
+}
+
+func ipTablesBootstrap(ipt IPTables, iptRestore IPTablesRestore, rules []IPTablesRule) error {
+	tablesRules, err := ipTablesCleanAndBuild(ipt, rules)
```
[the remark this answers is not in the document]
Comment (louiznk) [reply network/iptables.go:160]: I will do this, thanks

File: network/iptables.go [hunk 16]
```diff
@@ -135,24 +134,77 @@ func ipTablesRulesExist(ipt IPTables, rules []IPTablesRule) (bool, error) {
 	return true, nil
 }
 
+func ipTablesCleanAndBuild(ipt IPTables, rules []IPTablesRule) (IPTablesRestoreRules, error) {
+	tablesRules := IPTablesRestoreRules{}
```
[the remark this answers is not in the document]
Comment (louiznk) [reply network/iptables.go:138]: I will do this, thanks

File: network/iptables.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (louiznk) [reply network/iptables.go:164]: done @manuelbuil

File: network/iptables.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (louiznk) [reply network/iptables.go:139]: done @manuelbuil

File: network/iptables_restore.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (manuelbuil) [reply network/iptables_restore.go:56]: I personally would prefer to use the same functions for both ipv4 and ipv6 and change the arg accordingly. Therefore, I'd prefer `NewIPTablesRestoreWithProtocol(iptables.ProtocolIPv6)` and `NewIPTablesRestoreWithProtocol(iptables.ProtocolIPv4)` in the code. It feels to me more confusing if we use different functions for each protocol

**Comment by manuelbuil** [comment]
Tested and works

**Review by manuelbuil — APPROVED** [review]

**Pull request merged**